File: app/src/main/java/dev/ragnarok/fenrir/db/impl/KeysPersistStorage.java
package dev.ragnarok.fenrir.db.impl;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.content.ContentValues;
import android.database.Cursor;
import android.net.Uri;
import android.provider.BaseColumns;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.crypt.AesKeyPair;
import dev.ragnarok.fenrir.db.MessengerContentProvider;
import dev.ragnarok.fenrir.db.column.KeyColumns;
import dev.ragnarok.fenrir.db.interfaces.IKeysStorage;
import dev.ragnarok.fenrir.exception.DatabaseException;
import dev.ragnarok.fenrir.util.Optional;
import dev.ragnarok.fenrir.util.Utils;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Maybe;
import io.reactivex.rxjava3.core.Single;

class KeysPersistStorage extends AbsStorage implements IKeysStorage {

    KeysPersistStorage(@NonNull AppStorages context) {
        super(context);
    }

    private AesKeyPair map(Cursor cursor) {
        return new AesKeyPair()
                .setVersion(cursor.getInt(cursor.getColumnIndex(KeyColumns.VERSION)))
                .setPeerId(cursor.getInt(cursor.getColumnIndex(KeyColumns.PEER_ID)))
                .setSessionId(cursor.getLong(cursor.getColumnIndex(KeyColumns.SESSION_ID)))
                .setDate(cursor.getLong(cursor.getColumnIndex(KeyColumns.DATE)))
                .setStartMessageId(cursor.getInt(cursor.getColumnIndex(KeyColumns.START_SESSION_MESSAGE_ID)))
                .setEndMessageId(cursor.getInt(cursor.getColumnIndex(KeyColumns.END_SESSION_MESSAGE_ID)))
                .setHisAesKey(cursor.getString(cursor.getColumnIndex(KeyColumns.IN_KEY)))
                .setMyAesKey(cursor.getString(cursor.getColumnIndex(KeyColumns.OUT_KEY)));
    }

    @Override
    public Completable saveKeyPair(@NonNull AesKeyPair pair) {
        return Completable.create(e -> {
            AesKeyPair alreaadyExist = findKeyPairFor(pair.getAccountId(), pair.getSessionId())
                    .blockingGet();

            if (nonNull(alreaadyExist)) {
                e.onError(new DatabaseException("Key pair with the session ID is already in the database"));
                return;
            }

            ContentValues cv = new ContentValues();
            cv.put(KeyColumns.VERSION, pair.getVersion());
            cv.put(KeyColumns.PEER_ID, pair.getPeerId());
            cv.put(KeyColumns.SESSION_ID, pair.getSessionId());
            cv.put(KeyColumns.DATE, pair.getDate());
            cv.put(KeyColumns.START_SESSION_MESSAGE_ID, pair.getStartMessageId());
            cv.put(KeyColumns.END_SESSION_MESSAGE_ID, pair.getEndMessageId());
            cv.put(KeyColumns.OUT_KEY, pair.getMyAesKey());
            cv.put(KeyColumns.IN_KEY, pair.getHisAesKey());

            Uri uri = MessengerContentProvider.getKeysContentUriFor(pair.getAccountId());
            getContext().getContentResolver().insert(uri, cv);

            e.onComplete();
        });
    }

    @Override
    public Single<List<AesKeyPair>> getAll(int accountId) {
        return Single.create(e -> {
            Uri uri = MessengerContentProvider.getKeysContentUriFor(accountId);
            Cursor cursor = getContext().getContentResolver().query(uri, null, null, null, BaseColumns._ID);

            List<AesKeyPair> pairs = new ArrayList<>(Utils.safeCountOf(cursor));
            if (nonNull(cursor)) {
                while (cursor.moveToNext()) {
                    if (e.isDisposed()) {
                        break;
                    }

                    pairs.add(map(cursor).setAccountId(accountId));
                }

                cursor.close();
            }

            e.onSuccess(pairs);
        });
    }

    @Override
    public Single<List<AesKeyPair>> getKeys(int accountId, int peerId) {
        return Single.create(e -> {
            Uri uri = MessengerContentProvider.getKeysContentUriFor(accountId);
            Cursor cursor = getContext().getContentResolver()
                    .query(uri, null, KeyColumns.PEER_ID + " = ?", new String[]{String.valueOf(peerId)}, BaseColumns._ID);

            List<AesKeyPair> pairs = new ArrayList<>(Utils.safeCountOf(cursor));
            if (nonNull(cursor)) {
                while (cursor.moveToNext()) {
                    if (e.isDisposed()) {
                        break;
                    }

                    pairs.add(map(cursor).setAccountId(accountId));
                }
                cursor.close();
            }

            e.onSuccess(pairs);
        });
    }

    @Override
    public Single<Optional<AesKeyPair>> findLastKeyPair(int accountId, int peerId) {
        return Single.create(e -> {
            Uri uri = MessengerContentProvider.getKeysContentUriFor(accountId);
            Cursor cursor = getContext().getContentResolver()
                    .query(uri, null, KeyColumns.PEER_ID + " = ?",
                            new String[]{String.valueOf(peerId)}, BaseColumns._ID + " DESC LIMIT 1");

            AesKeyPair pair = null;
            if (nonNull(cursor)) {
                if (cursor.moveToNext()) {
                    pair = map(cursor).setAccountId(accountId);
                }

                cursor.close();
            }

            e.onSuccess(Optional.wrap(pair));
        });
    }

    @Override
    public Maybe<AesKeyPair> findKeyPairFor(int accountId, long sessionId) {
        return Maybe.create(e -> {
            Uri uri = MessengerContentProvider.getKeysContentUriFor(accountId);
            Cursor cursor = getContext().getContentResolver()
                    .query(uri, null, KeyColumns.SESSION_ID + " = ?",
                            new String[]{String.valueOf(sessionId)}, null);

            AesKeyPair pair = null;
            if (nonNull(cursor)) {
                if (cursor.moveToNext()) {
                    pair = map(cursor).setAccountId(accountId);
                }

                cursor.close();
            }

            if (nonNull(pair)) {
                e.onSuccess(pair);
            }

            e.onComplete();
        });
    }


    @Override
    public Completable deleteAll(int accountId) {
        return Completable.create(e -> {
            Uri uri = MessengerContentProvider.getKeysContentUriFor(accountId);
            getContext().getContentResolver().delete(uri, null, null);
            e.onComplete();
        });
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/db/impl/KeysPersistStorage.kt
package dev.ragnarok.fenrir.db.impl

import android.content.ContentValues
import android.database.Cursor
import android.provider.BaseColumns
import dev.ragnarok.fenrir.crypt.AesKeyPair
import dev.ragnarok.fenrir.db.MessengerContentProvider.Companion.getKeysContentUriFor
import dev.ragnarok.fenrir.db.column.KeyColumns
import dev.ragnarok.fenrir.db.interfaces.IKeysStorage
import dev.ragnarok.fenrir.exception.DatabaseException
import dev.ragnarok.fenrir.getInt
import dev.ragnarok.fenrir.getLong
import dev.ragnarok.fenrir.getString
import dev.ragnarok.fenrir.util.Optional
import dev.ragnarok.fenrir.util.Optional.Companion.wrap
import dev.ragnarok.fenrir.util.Utils.safeCountOf
import io.reactivex.rxjava3.core.*

internal class KeysPersistStorage(context: AppStorages) : AbsStorage(context), IKeysStorage {
    private fun map(cursor: Cursor): AesKeyPair {
        return AesKeyPair()
            .setVersion(cursor.getInt(KeyColumns.VERSION))
            .setPeerId(cursor.getInt(KeyColumns.PEER_ID))
            .setSessionId(cursor.getLong(KeyColumns.SESSION_ID))
            .setDate(cursor.getLong(KeyColumns.DATE))
            .setStartMessageId(cursor.getInt(KeyColumns.START_SESSION_MESSAGE_ID))
            .setEndMessageId(cursor.getInt(KeyColumns.END_SESSION_MESSAGE_ID))
            .setHisAesKey(cursor.getString(KeyColumns.IN_KEY))
            .setMyAesKey(cursor.getString(KeyColumns.OUT_KEY))
    }

    override fun saveKeyPair(pair: AesKeyPair): Completable {
        return Completable.create { e: CompletableEmitter ->
            val alreaadyExist = findKeyPairFor(pair.accountId, pair.sessionId)
                .blockingGet()
            if (alreaadyExist != null) {
                e.onError(DatabaseException("Key pair with the session ID is already in the database"))
                return@create
            }
            val cv = ContentValues()
            cv.put(KeyColumns.VERSION, pair.version)
            cv.put(KeyColumns.PEER_ID, pair.peerId)
            cv.put(KeyColumns.SESSION_ID, pair.sessionId)
            cv.put(KeyColumns.DATE, pair.date)
            cv.put(KeyColumns.START_SESSION_MESSAGE_ID, pair.startMessageId)
            cv.put(KeyColumns.END_SESSION_MESSAGE_ID, pair.endMessageId)
            cv.put(KeyColumns.OUT_KEY, pair.myAesKey)
            cv.put(KeyColumns.IN_KEY, pair.hisAesKey)
            val uri = getKeysContentUriFor(pair.accountId)
            context.contentResolver.insert(uri, cv)
            e.onComplete()
        }
    }

    override fun getAll(accountId: Int): Single<List<AesKeyPair>> {
        return Single.create { e: SingleEmitter<List<AesKeyPair>> ->
            val uri = getKeysContentUriFor(accountId)
            val cursor = context.contentResolver.query(uri, null, null, null, BaseColumns._ID)
            val pairs: MutableList<AesKeyPair> = ArrayList(safeCountOf(cursor))
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    if (e.isDisposed) {
                        break
                    }
                    pairs.add(map(cursor).setAccountId(accountId))
                }
                cursor.close()
            }
            e.onSuccess(pairs)
        }
    }

    override fun getKeys(accountId: Int, peerId: Int): Single<List<AesKeyPair>> {
        return Single.create { e: SingleEmitter<List<AesKeyPair>> ->
            val uri = getKeysContentUriFor(accountId)
            val cursor = context.contentResolver
                .query(
                    uri,
                    null,
                    KeyColumns.PEER_ID + " = ?",
                    arrayOf(peerId.toString()),
                    BaseColumns._ID
                )
            val pairs: MutableList<AesKeyPair> = ArrayList(safeCountOf(cursor))
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    if (e.isDisposed) {
                        break
                    }
                    pairs.add(map(cursor).setAccountId(accountId))
                }
                cursor.close()
            }
            e.onSuccess(pairs)
        }
    }

    override fun findLastKeyPair(accountId: Int, peerId: Int): Single<Optional<AesKeyPair>> {
        return Single.create { e: SingleEmitter<Optional<AesKeyPair>> ->
            val uri = getKeysContentUriFor(accountId)
            val cursor = context.contentResolver
                .query(
                    uri,
                    null,
                    KeyColumns.PEER_ID + " = ?",
                    arrayOf(peerId.toString()),
                    BaseColumns._ID + " DESC LIMIT 1"
                )
            var pair: AesKeyPair? = null
            if (cursor != null) {
                if (cursor.moveToNext()) {
                    pair = map(cursor).setAccountId(accountId)
                }
                cursor.close()
            }
            e.onSuccess(wrap(pair))
        }
    }

    override fun findKeyPairFor(accountId: Int, sessionId: Long): Maybe<AesKeyPair> {
        return Maybe.create { e: MaybeEmitter<AesKeyPair> ->
            val uri = getKeysContentUriFor(accountId)
            val cursor = context.contentResolver
                .query(
                    uri,
                    null,
                    KeyColumns.SESSION_ID + " = ?",
                    arrayOf(sessionId.toString()),
                    null
                )
            var pair: AesKeyPair? = null
            if (cursor != null) {
                if (cursor.moveToNext()) {
                    pair = map(cursor).setAccountId(accountId)
                }
                cursor.close()
            }
            if (pair != null) {
                e.onSuccess(pair)
            }
            e.onComplete()
        }
    }

    override fun deleteAll(accountId: Int): Completable {
        return Completable.create { e: CompletableEmitter ->
            val uri = getKeysContentUriFor(accountId)
            context.contentResolver.delete(uri, null, null)
            e.onComplete()
        }
    }
}